File: go/jewelmusic/http_client.go
package jewelmusic

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// HTTPClient handles HTTP communication with the JewelMusic API
type HTTPClient struct {
	baseURL    string
	apiKey     string
	httpClient *http.Client
	userAgent  string
}

// NewHTTPClient creates a new HTTP client
func NewHTTPClient(baseURL, apiKey string, httpClient *http.Client) *HTTPClient {
	if httpClient == nil {
		httpClient = &http.Client{
			Timeout: 30 * time.Second,
		}
	}

	return &HTTPClient{
		baseURL:    strings.TrimSuffix(baseURL, "/"),
		apiKey:     apiKey,
		httpClient: httpClient,
		userAgent:  "JewelMusic-Go-SDK/1.0.0",
	}
}

// APIResponse represents a standard API response
type APIResponse struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data"`
	Meta    struct {
		Timestamp string `json:"timestamp"`
		RequestID string `json:"requestId"`
		RateLimit struct {
			Limit     int `json:"limit"`
			Remaining int `json:"remaining"`
			Reset     int `json:"reset"`
		} `json:"rateLimit"`
	} `json:"meta"`
	Error *APIError `json:"error,omitempty"`
}

// APIError represents an API error response
type APIError struct {
	Code    string                 `json:"code"`
	Message string                 `json:"message"`
	Details map[string]interface{} `json:"details"`
}

// Error implements the error interface
func (e *APIError) Error() string {
	return fmt.Sprintf("API Error %s: %s", e.Code, e.Message)
}

// makeRequest performs an HTTP request with retries and error handling
func (c *Client) makeRequest(ctx context.Context, method, path string, body interface{}, result interface{}) error {
	// Build URL
	url := c.baseURL + "/v1" + path

	// Prepare request body
	var bodyReader io.Reader
	var contentType string

	if body != nil {
		switch v := body.(type) {
		case *multipart.Writer:
			// For file uploads
			bodyReader = strings.NewReader("") // Placeholder, actual implementation would be different
			contentType = v.FormDataContentType()
		default:
			// JSON body
			jsonBody, err := json.Marshal(body)
			if err != nil {
				return fmt.Errorf("failed to marshal request body: %w", err)
			}
			bodyReader = bytes.NewReader(jsonBody)
			contentType = "application/json"
		}
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("User-Agent", c.httpClient.(*HTTPClient).userAgent)
	req.Header.Set("Accept", "application/json")
	
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}

	// Perform request with retries
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response body: %w", err)
	}

	// Parse response
	var apiResp APIResponse
	if err := json.Unmarshal(respBody, &apiResp); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	// Handle errors
	if resp.StatusCode >= 400 {
		if apiResp.Error != nil {
			return apiResp.Error
		}
		return fmt.Errorf("API request failed with status %d", resp.StatusCode)
	}

	// Extract data if result is provided
	if result != nil && apiResp.Data != nil {
		dataBytes, err := json.Marshal(apiResp.Data)
		if err != nil {
			return fmt.Errorf("failed to marshal response data: %w", err)
		}

		if err := json.Unmarshal(dataBytes, result); err != nil {
			return fmt.Errorf("failed to unmarshal response data: %w", err)
		}
	}

	return nil
}

// Get performs a GET request
func (c *Client) Get(ctx context.Context, path string, params map[string]string, result interface{}) error {
	if params != nil && len(params) > 0 {
		query := url.Values{}
		for k, v := range params {
			query.Add(k, v)
		}
		path += "?" + query.Encode()
	}
	return c.makeRequest(ctx, "GET", path, nil, result)
}

// Post performs a POST request
func (c *Client) Post(ctx context.Context, path string, body interface{}, result interface{}) error {
	return c.makeRequest(ctx, "POST", path, body, result)
}

// Put performs a PUT request
func (c *Client) Put(ctx context.Context, path string, body interface{}, result interface{}) error {
	return c.makeRequest(ctx, "PUT", path, body, result)
}

// Delete performs a DELETE request
func (c *Client) Delete(ctx context.Context, path string, result interface{}) error {
	return c.makeRequest(ctx, "DELETE", path, nil, result)
}

// UploadFile uploads a file with metadata
func (c *Client) UploadFile(ctx context.Context, path string, file io.Reader, filename string, metadata map[string]string) (*APIResponse, error) {
	var buf bytes.Buffer
	writer := multipart.NewWriter(&buf)

	// Add metadata fields
	for key, value := range metadata {
		if err := writer.WriteField(key, value); err != nil {
			return nil, fmt.Errorf("failed to write field %s: %w", key, err)
		}
	}

	// Add file
	part, err := writer.CreateFormFile("file", filename)
	if err != nil {
		return nil, fmt.Errorf("failed to create form file: %w", err)
	}

	if _, err := io.Copy(part, file); err != nil {
		return nil, fmt.Errorf("failed to copy file: %w", err)
	}

	if err := writer.Close(); err != nil {
		return nil, fmt.Errorf("failed to close writer: %w", err)
	}

	// Create request
	url := c.baseURL + "/v1" + path
	req, err := http.NewRequestWithContext(ctx, "POST", url, &buf)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("User-Agent", c.httpClient.(*HTTPClient).userAgent)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	// Perform request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("upload failed: %w", err)
	}
	defer resp.Body.Close()

	// Parse response
	var apiResp APIResponse
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if err := json.Unmarshal(respBody, &apiResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	if resp.StatusCode >= 400 {
		if apiResp.Error != nil {
			return nil, apiResp.Error
		}
		return nil, fmt.Errorf("upload failed with status %d", resp.StatusCode)
	}

	return &apiResp, nil
}

File: python/jewelmusic_sdk/utils/http_client.py
"""
HTTP Client for JewelMusic Python SDK

Handles all HTTP communication with the JewelMusic API including
authentication, retries, rate limiting, and error handling.
"""

import asyncio
import json
import time
from typing import Optional, Dict, Any, Union, BinaryIO, Tuple
from urllib.parse import urljoin, urlencode
import aiohttp
import aiofiles
from ..exceptions import create_error_from_response, NetworkError, ConfigurationError


class HttpClient:
    """
    HTTP client for JewelMusic API communication.
    
    Handles authentication, request/response processing, retries,
    rate limiting, and error handling for all API requests.
    """
    
    def __init__(
        self,
        api_key: str,
        base_url: str,
        api_version: str = 'v1',
        timeout: float = 30.0,
        max_retries: int = 3,
        retry_delay: float = 1.0,
        user_agent: Optional[str] = None,
        **kwargs
    ):
        """
        Initialize the HTTP client.
        
        Args:
            api_key: JewelMusic API key
            base_url: Base URL for API endpoints
            api_version: API version to use
            timeout: Request timeout in seconds
            max_retries: Maximum retry attempts
            retry_delay: Initial retry delay in seconds
            user_agent: Custom user agent string
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.api_version = api_version
        self.timeout = aiohttp.ClientTimeout(total=timeout)
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        
        # Build user agent
        if user_agent is None:
            user_agent = f"JewelMusic-Python-SDK/1.0.0"
        self.user_agent = user_agent
        
        # Default headers
        self.default_headers = {
            'Authorization': f'Bearer {api_key}',
            'User-Agent': user_agent,
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
        
        # HTTP session (created lazily)
        self._session: Optional[aiohttp.ClientSession] = None
        
        # Rate limiting state
        self._rate_limit_remaining = None
        self._rate_limit_reset = None
        self._rate_limit_limit = None
    
    @property
    def session(self) -> aiohttp.ClientSession:
        """Get or create the HTTP session."""
        if self._session is None or self._session.closed:
            connector = aiohttp.TCPConnector(
                limit=100,  # Connection pool size
                limit_per_host=30,
                ttl_dns_cache=300,
                use_dns_cache=True,
            )
            
            self._session = aiohttp.ClientSession(
                connector=connector,
                timeout=self.timeout,
                headers=self.default_headers,
                raise_for_status=False  # We'll handle status codes manually
            )
        
        return self._session
    
    def _build_url(self, path: str) -> str:
        """Build complete URL from path."""
        if path.startswith('/'):
            path = path[1:]
        
        # Add API version prefix if not already present
        if not path.startswith(f'{self.api_version}/'):
            path = f'{self.api_version}/{path}'
        
        return urljoin(f'{self.base_url}/', path)
    
    async def _handle_rate_limiting(self) -> None:
        """Handle rate limiting by waiting if necessary."""
        if (self._rate_limit_remaining is not None and 
            self._rate_limit_remaining <= 5 and  # Close to limit
            self._rate_limit_reset is not None):
            
            # Calculate wait time
            wait_time = max(0, self._rate_limit_reset - time.time())
            if wait_time > 0:
                await asyncio.sleep(min(wait_time, 60))  # Max 1 minute wait
    
    def _update_rate_limit_info(self, headers: Dict[str, str]) -> None:
        """Update rate limit information from response headers."""
        try:
            if 'X-RateLimit-Remaining' in headers:
                self._rate_limit_remaining = int(headers['X-RateLimit-Remaining'])
            if 'X-RateLimit-Limit' in headers:
                self._rate_limit_limit = int(headers['X-RateLimit-Limit'])
            if 'X-RateLimit-Reset' in headers:
                self._rate_limit_reset = int(headers['X-RateLimit-Reset'])
        except (ValueError, TypeError):
            # Ignore invalid rate limit headers
            pass
    
    async def _make_request(
        self,
        method: str,
        url: str,
        headers: Optional[Dict[str, str]] = None,
        data: Optional[Union[str, bytes, aiohttp.FormData]] = None,
        params: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> aiohttp.ClientResponse:
        """
        Make HTTP request with retries and error handling.
        
        Args:
            method: HTTP method
            url: Request URL
            headers: Additional headers
            data: Request body data
            params: Query parameters
            
        Returns:
            HTTP response
            
        Raises:
            NetworkError: For network-related errors
            Various API errors based on response status
        """
        # Check rate limiting
        await self._handle_rate_limiting()
        
        # Merge headers
        request_headers = self.default_headers.copy()
        if headers:
            request_headers.update(headers)
        
        # Build query string
        if params:
            # Filter out None values
            params = {k: v for k, v in params.items() if v is not None}
            if params:
                url = f"{url}?{urlencode(params, doseq=True)}"
        
        last_exception = None
        
        for attempt in range(self.max_retries + 1):
            try:
                async with self.session.request(
                    method=method,
                    url=url,
                    headers=request_headers,
                    data=data,
                    **kwargs
                ) as response:
                    # Update rate limit info
                    self._update_rate_limit_info(dict(response.headers))
                    
                    # Handle successful responses
                    if 200 <= response.status < 300:
                        return response
                    
                    # Handle rate limiting
                    if response.status == 429:
                        if attempt < self.max_retries:
                            retry_after = response.headers.get('Retry-After')
                            if retry_after:
                                wait_time = min(float(retry_after), 60)
                            else:
                                wait_time = min(self.retry_delay * (2 ** attempt), 60)
                            
                            await asyncio.sleep(wait_time)
                            continue
                    
                    # Handle server errors with retry
                    if response.status >= 500 and attempt < self.max_retries:
                        wait_time = self.retry_delay * (2 ** attempt)
                        await asyncio.sleep(wait_time)
                        continue
                    
                    # Parse error response
                    try:
                        error_data = await response.json()
                    except (json.JSONDecodeError, aiohttp.ContentTypeError):
                        error_data = {
                            'error': {
                                'message': f'HTTP {response.status}: {response.reason}',
                                'details': {}
                            }
                        }
                    
                    # Create and raise appropriate exception
                    raise create_error_from_response(error_data, response.status)
            
            except aiohttp.ClientError as e:
                last_exception = NetworkError(f"Network error: {str(e)}")
                
                if attempt < self.max_retries:
                    wait_time = self.retry_delay * (2 ** attempt)
                    await asyncio.sleep(wait_time)
                    continue
                else:
                    break
            
            except asyncio.TimeoutError:
                last_exception = NetworkError("Request timeout")
                
                if attempt < self.max_retries:
                    wait_time = self.retry_delay * (2 ** attempt)
                    await asyncio.sleep(wait_time)
                    continue
                else:
                    break
        
        # If we get here, all retries failed
        if last_exception:
            raise last_exception
        else:
            raise NetworkError("Request failed after all retries")
    
    async def request(
        self,
        method: str,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Make a JSON API request.
        
        Args:
            method: HTTP method
            path: API endpoint path
            json_data: JSON request data
            params: Query parameters
            headers: Additional headers
            
        Returns:
            Parsed JSON response data
        """
        url = self._build_url(path)
        
        # Prepare request data
        data = None
        if json_data is not None:
            data = json.dumps(json_data, separators=(',', ':'))
        
        response = await self._make_request(
            method=method,
            url=url,
            headers=headers,
            data=data,
            params=params,
            **kwargs
        )
        
        # Parse JSON response
        try:
            response_data = await response.json()
        except (json.JSONDecodeError, aiohttp.ContentTypeError):
            response_data = {}
        
        return response_data
    
    async def upload_file(
        self,
        path: str,
        file_data: Union[bytes, BinaryIO],
        metadata: Optional[Dict[str, Any]] = None,
        filename: Optional[str] = None,
        content_type: Optional[str] = None,
        progress_callback: Optional[callable] = None
    ) -> Dict[str, Any]:
        """
        Upload a file to the API.
        
        Args:
            path: API endpoint path
            file_data: File data (bytes or file-like object)
            metadata: Additional form data
            filename: Name of the file
            content_type: MIME type of the file
            progress_callback: Optional progress callback function
            
        Returns:
            Parsed JSON response data
        """
        url = self._build_url(path)
        
        # Create multipart form data
        form_data = aiohttp.FormData()
        
        # Add metadata fields
        if metadata:
            for key, value in metadata.items():
                if value is not None:
                    if isinstance(value, (dict, list)):
                        form_data.add_field(key, json.dumps(value))
                    else:
                        form_data.add_field(key, str(value))
        
        # Add file data
        if isinstance(file_data, bytes):
            form_data.add_field(
                'file',
                file_data,
                filename=filename or 'upload',
                content_type=content_type or 'application/octet-stream'
            )
        else:
            # File-like object
            if hasattr(file_data, 'read'):
                content = file_data.read()
                if hasattr(file_data, 'seek'):
                    file_data.seek(0)  # Reset file pointer
                
                form_data.add_field(
                    'file',
                    content,
                    filename=filename or getattr(file_data, 'name', 'upload'),
                    content_type=content_type or 'application/octet-stream'
                )
            else:
                raise ValueError("file_data must be bytes or file-like object")
        
        # Remove Content-Type header for multipart uploads
        headers = self.default_headers.copy()
        headers.pop('Content-Type', None)
        
        response = await self._make_request(
            method='POST',
            url=url,
            headers=headers,
            data=form_data
        )
        
        # Parse JSON response
        try:
            response_data = await response.json()
        except (json.JSONDecodeError, aiohttp.ContentTypeError):
            response_data = {}
        
        return response_data
    
    async def download_file(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        progress_callback: Optional[callable] = None
    ) -> bytes:
        """
        Download a file from the API.
        
        Args:
            path: API endpoint path
            params: Query parameters
            progress_callback: Optional progress callback function
            
        Returns:
            File content as bytes
        """
        url = self._build_url(path)
        
        response = await self._make_request(
            method='GET',
            url=url,
            params=params
        )
        
        # Read file content
        content = await response.read()
        
        if progress_callback:
            progress_callback(len(content), len(content))
        
        return content
    
    async def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a GET request."""
        return await self.request('GET', path, params=params)
    
    async def post(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make a POST request."""
        return await self.request('POST', path, json_data=json_data, params=params)
    
    async def put(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make a PUT request."""
        return await self.request('PUT', path, json_data=json_data, params=params)
    
    async def patch(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make a PATCH request."""
        return await self.request('PATCH', path, json_data=json_data, params=params)
    
    async def delete(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make a DELETE request."""
        return await self.request('DELETE', path, params=params)
    
    async def close(self) -> None:
        """Close the HTTP session and cleanup resources."""
        if self._session and not self._session.closed:
            await self._session.close()